### ISSUE TITLE
Exclude URLs from Tracing

### ISSUE THREAD

**Issue by jakob-o** [open-telemetry/opentelemetry-java#1552]
**Is your feature request related to a problem? Please describe.**
As already mentioned here https://github.com/open-telemetry/opentelemetry-specification/issues/173 I'd like to be able to exclude or sample a list of URLs / URL-Patterns from instrumentation. In my case particularly to avoid generating many events from health- and liveness-checks.

**Describe the solution you'd like**
If I understood correctly in the python implementation the Context flag `suppress_instrumentation` indicates that the current context should not be instrumented.
What could be a proper implementation in the Java API / SDK?

Would it be suitable to implement it as a flag in [TraceFlags](https://github.com/open-telemetry/opentelemetry-java/blob/a08dd9fbe4a98f34aec3618475905fa55fece87e/api/src/main/java/io/opentelemetry/trace/TraceFlags.java) and make [SpanBuilderSdk#startSpan](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java#L211) return a DefaultSpan as it would if the sampling decision was negative?
This flag could maybe set in [HttpServerTracer](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1b9b8b49b6413039493ac9d63a425ed0a7d572f4/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpServerTracer.java#L74) depending on some configurable input?

**Describe alternatives you've considered**
Unfortunately I am not very familiar with the OTEL architecture and any hint to where certain information and or decisions should be made or where I am able to get a high level view of the Java SDK is highly appreciated.

**Comment by jkwatson** [comment]
Hi @jakob-o . Thanks for this request.

If you're using the auto-instrumentation agent, then this issue would probably belong in that repo: https://github.com/open-telemetry/opentelemetry-java-instrumentation

Otherwise, the instrumentation you apply is totally up to you, unless I'm misunderstanding the request.

**Comment by jakob-o** [comment]
Hey @jkwatson,

maybe you are right and this belongs to the instrumentation repository, but if I understood correctly in the API discussion it might be beneficial to have a general option such as a context flag to disable tracing centrally which can be set anywhere for example in a filter which would lead to generating noop-spans. Otherwise every (auto) instrumentation implementation would require to manually check somewhere (?) whether it should take place for the current request.

Does this make sense?

**Comment by jakob-o** [comment]
Sorry, the discussion related to the referenced issue was actually taking place here https://github.com/open-telemetry/opentelemetry-specification/issues/530

**Comment by iNikem** [comment]
@jkwatson I think there was an idea to allow for `Tracer` to check for some flag/configuration and return non-recording/invalid span from `startSpan`

**Comment by jkwatson** [comment]
Hmm. I'm trying to figure out if the instrumentation needs a feature to do this. If the instrumentation is making the decision, based on instrumentation configuration, can't the instrumentation choose to not create the span at all in the first place?

**Comment by jakob-o** [comment]
I think the issue is, that instrumentation might take place at several layers and if the HTTP instrumentation does decide to not create a span the JDBC instrumentation will not know this, unless it implements the same logic and will create a new span itself.
This would not be necessary if the tracer itself has means to disable it centrally.

**Comment by iNikem** [comment]
The problem is not local to one instrumentation making a decision. E.g. my servlet instrumentation can detect excluded url pattern and create non-recording span. But after that there is not point in creating proper child spans for this SERVER span, is it? All instrumentations now must check if parent span is non-recording and probably skip creating any spans at all.

**Comment by jkwatson** [comment]
> The problem is not local to one instrumentation making a decision. E.g. my servlet instrumentation can detect excluded url pattern and create non-recording span. But after that there is not point in creating proper child spans for this SERVER span, is it? All instrumentations now must check if parent span is non-recording and probably skip creating any spans at all.

This feels like a separate issue (not unimportant!). Don't we have this issue today with all kinds of parents? Should the SDK not create child spans at all if the parent is non-recording?

Maybe I'm getting hung up on the name of this Issue. This seems like a much broader issue than excluding URLs from tracing in HTTP server spans. :)

**Comment by jakob-o** [comment]
The decision making should surely be implemented in the instrumentation. My question is: How do i correctly disable recording for the span / trace?
Does this feature require to make every instrumentation check the span, or would a central way to disable it as discussed here https://github.com/open-telemetry/opentelemetry-specification/issues/530 make more sense?

**Comment by jakob-o** [comment]
@jkwatson would you like me to rename / rephrase the issue to something like `Option to disable tracing centrally`?

**Comment by trask** [comment]
It sounds like this is proposing to implement https://github.com/open-telemetry/opentelemetry-specification/issues/530? I'm all for this, but would be great if we can get this into the spec.

**Comment by jkwatson** [comment]
That spec issue is marked as after-ga, so I'm going to follow suit for this one, for the nonce. If it ends up getting fast-tracked, we can change it here, as well.

**Comment by Oberon00** [comment]
>  Should the SDK not create child spans at all if the parent is non-recording?

I think it already implements that 😃  It's called ParentOrElse sampler.

**Comment by jakob-o** [comment]
@jkwatson Would you recommend opening another issue regarding the URL exclusion in https://github.com/open-telemetry/opentelemetry-java-instrumentation and reference this issue?

@Oberon00 Would it be sufficient then to create a non-recording span in the HTTP instrumentation and create / enable this sampler? If so, do you know how and where would I do this appropriately?

**Comment by jkwatson** [comment]
yes, I think having an issue open for the instrumentation itself would be great. 

There are configuration options in the agent to set the sampling setup; Configuration options are here: https://github.com/open-telemetry/opentelemetry-java-instrumentation#trace-config

**Comment by anuraaga** [comment]
> This feels like a separate issue (not unimportant!). Don't we have this issue today with all kinds of parents? Should the SDK not create child spans at all if the parent is non-recording?

Just to clarify this point, currently we create IDs for all spans, even non-recording ones. I think this is to allow the IDs to be injected into logs, for example. If the spans weren't valid, they wouldn't be injected nor would children be created.

**Comment by iNikem** [comment]
I don't think we should create spans nor spanIds for "ignored" requests. I don't know if the correct way to that is via non-recording spans or with `TraceContext.sampled=false`

**Comment by jaisonsteephen** [comment]
@jakob-o , by exporting OTEL_JAVAAGENT_ENABLED=false, you can disable trace in java

**Comment by kmendo** [comment]
> @jakob-o , by exporting OTEL_JAVAAGENT_ENABLED=false, you can disable trace in java

Where do you get this from? there is no reference for this flag in the code, nor it is documented.

**Comment by jkwatson** [comment]
> > @jakob-o , by exporting OTEL_JAVAAGENT_ENABLED=false, you can disable trace in java
> 
> Where do you get this from? there is no reference for this flag in the code, nor it is documented.

this is in the instrumentation codebase, not in this one.
